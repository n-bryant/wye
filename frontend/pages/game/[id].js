import React from "react";
import PropTypes from "prop-types";
import { useRouter } from "next/router";

import get from "lodash.get";
import gql from "graphql-tag";
import { useQuery } from "@apollo/react-hooks";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import LoadingState from "../../src/components/LoadingState";
import GameDetails from "../../src/components/Game/GameDetails";

import styles from "./[id].styles";

// game query
export const GET_GAME_QUERY = gql`
  query GET_GAME_QUERY($gameId: ID!) {
    game(gameId: $gameId) {
      details(gameId: $gameId) {
        id
        name
        releaseDate
        shortDescription
        headerImageUrl
        backgroundImageUrl
        price {
          freeToPlay
          onSale
          discountPercentage
          initialFormatted
          finalFormatted
        }
        controllerSupport
        developers
        publishers
        website
        platforms
        requirements {
          minimum
          recommended
        }
        metacritic {
          score
          reviewsPageUrl
        }
        categories {
          description
        }
        genres {
          description
        }
        highlightedVideos {
          id
          title
          thumbnailUrl
          fullsizeUrl
        }
        screenshots {
          id
          thumbnailUrl
          fullsizeUrl
        }
        videos {
          id
          title
          thumbnailUrl
          fullsizeUrl
        }
      }
      articles(gameId: $gameId) {
        url
        printDate
        author
        title
        contents
      }
    }
  }
`;

/**
 * renders a game page
 */
export const Game = props => {
  const classnames = Game.classnames(props);
  const router = useRouter();

  const { data, loading, error } = useQuery(GET_GAME_QUERY, {
    variables: {
      gameId: router.query.id
    }
  });
  const details = get(data, ["game", "details"]);
  const articles = get(data, ["game", "articles"]);

  if (loading) {
    // game query loading state
    return (
      <div className={classnames.root()}>
        <LoadingState />
      </div>
    );
  }

  if (error) {
    // game query error state
  }

  return (
    <div className={classnames.root()}>
      {details && articles && <GameDetails data={{ details, articles }} />}
    </div>
  );
};
Game.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Game`);
Game.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string
  })
};
Game.defaultProps = {
  classes: {}
};

export default withStyles(styles)(Game);
