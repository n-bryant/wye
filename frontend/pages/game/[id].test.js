import { MockedProvider } from "@apollo/react-testing";
import { Game, GET_GAME_QUERY } from "./[id]";

jest.mock("next/router", () => ({
  useRouter: () => ({
    query: {
      id: "1"
    }
  })
}));

describe("Game", () => {
  const props = {
    classes: {
      root: "root"
    }
  };

  const mocks = [
    {
      request: {
        query: GET_GAME_QUERY,
        variables: {
          gameId: "1"
        }
      },
      result: {
        data: {
          details: {},
          articles: []
        }
      }
    }
  ];

  it("should render successfully", () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Game {...props} />
      </MockedProvider>
    );
    expect(toJson(wrapper.find(Game))).toMatchSnapshot();
  });
});
