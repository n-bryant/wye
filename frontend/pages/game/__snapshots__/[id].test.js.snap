// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render successfully 1`] = `
<Game
  classes={
    Object {
      "root": "root",
    }
  }
>
  <div
    className="wye-Game root"
  >
    <WithStyles(LoadingState)
      classes={Object {}}
    >
      <LoadingState
        classes={
          Object {
            "progressIndicator": "LoadingState-progressIndicator-2",
            "root": "LoadingState-root-1",
          }
        }
      >
        <div
          className="wye-LoadingState LoadingState-root-1"
        >
          <WithStyles(ForwardRef(CircularProgress))
            className="wye-LoadingState__progressIndicator LoadingState-progressIndicator-2"
          >
            <ForwardRef(CircularProgress)
              className="wye-LoadingState__progressIndicator LoadingState-progressIndicator-2"
              classes={
                Object {
                  "circle": "MuiCircularProgress-circle",
                  "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                  "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                  "circleStatic": "MuiCircularProgress-circleStatic",
                  "colorPrimary": "MuiCircularProgress-colorPrimary",
                  "colorSecondary": "MuiCircularProgress-colorSecondary",
                  "indeterminate": "MuiCircularProgress-indeterminate",
                  "root": "MuiCircularProgress-root",
                  "static": "MuiCircularProgress-static",
                  "svg": "MuiCircularProgress-svg",
                }
              }
            >
              <div
                className="MuiCircularProgress-root wye-LoadingState__progressIndicator LoadingState-progressIndicator-2 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style={
                  Object {
                    "height": 40,
                    "width": 40,
                  }
                }
              >
                <svg
                  className="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx={44}
                    cy={44}
                    fill="none"
                    r={20.2}
                    strokeWidth={3.6}
                    style={Object {}}
                  />
                </svg>
              </div>
            </ForwardRef(CircularProgress)>
          </WithStyles(ForwardRef(CircularProgress))>
        </div>
      </LoadingState>
    </WithStyles(LoadingState)>
  </div>
</Game>
`;
